### PR TITLE
[DTensor] Add DTensor redistribute fwd/bwd datatype conversion to enable SimpleFSDP mixed precision training

### PR DESCRIPTION
**Pull request by ruisizhang123**
As titled, this pr adds additional `forward_dtype` and `backward_dtype` conversion in DTensor `redistribute` API to enable SimpleFSDP's mixed precision training. 

In this forward pass, the DTensor can be configured to be cast to `forward_dtype`; in the backward pass, the DTensor can be configured to be cast to `backward_dtype`.

1. **Correctness**: The end-to-end SimpleFSDP mixed precision training integration has been proved to work properly in the PR from this fork: https://github.com/tianyu-l/pytorch_intern24/pull/20. We are now migrating the code to official PyTorch DTensor.

2. **Example Usage**: There is an example in TorchTian's SimpleFSDP implementation: https://github.com/pytorch/torchtitan/pull/1060. 

In the example below, a DTensor `x` is all-gather'ed along the `self.compute_placements`, with datatype cast to `self.param_dtype`. In the backward pass, additionally, the computed gradients are reduce-scatter'ed along the `self.grad_placements`, with datatype cast to `self.reduce_dtype`.

```python
output = x.redistribute(
        placements=self.compute_placements,
        forward_dtype=self.param_dtype,
        backward_dtype=self.reduce_dtype,
).to_local(grad_placements=self.grad_placements)
```

Under the hood, in `class Redistribute(torch.autograd.Function):`, the `forward` function first takes `x`'s local tensor, convert it to `forward_dtype`, before all-gather `x`. 

The `backward` function take `grad_output` and convert it to `backward_dtype`, before reduce-scatter `grad_output`.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
could you add a bit more context to the PR description? 
1) concise example / pseudocode for how this would be used (by fsdp)
2) link to fsdp  code with the full details

File: test/distributed/tensor/test_redistribute.py [hunk 1]
```diff
@@ -24,6 +24,14 @@ class RedistributeTest(DTensorTestBase):
     def world_size(self):
         return 4
 
+    @property
+    def forward_dtype(self):
+        return torch.bfloat16
+
+    @property
+    def backward_dtype(self):
+        return torch.float32
```
Comment (tianyu-l) [inline test/distributed/tensor/test_redistribute.py:33]: Since they are used only in two test functions, not all of them. Let's not set them as global properties. We can just put them in the relevant tests.

File: torch/distributed/tensor/_redistribute.py [hunk 2]
```diff
@@ -320,10 +338,29 @@ def forward(  # type: ignore[override]
     @staticmethod
     def backward(ctx, grad_output: "dtensor.DTensor"):  # type: ignore[override]
         previous_spec = ctx.current_spec
-        current_spec = grad_output._spec
         async_op = ctx.async_op
+        backward_dtype = ctx.backward_dtype
+
+        if backward_dtype is not None:
+            local_tensor = grad_output._local_tensor.to(dtype=backward_dtype)
+            current_spec = DTensorSpec(
```
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/tensor/_redistribute.py:346]: @wconstab 
This is a DTensor redistribute API change, which should not has binding with how it's used.
In SimpleFSDP, the redistribute is called on model parameters, not activations. There is no mixed precision for activations, only for parameters; activations are always with low precisions (first time obtained from the low precision embedding layer on int-type token inputs).

For example usage, please see https://github.com/pytorch/torchtitan/pull/1060

[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:346]: yea, agreed the API change should be 'general' and its ok to not put simple-fsdp example into the code-docs.  But what I am looking for is a more intuitive explanation of the API behavior.

1) the DTensor starts off with some `initial dtype`
2) redistribute adds a 'forward' and 'backward' dtype, but leaves the DTensor in 'initial dtype' (?)
3) during forward, the _local_tensor is converted from 'initial dtype' to 'forward_dtype' (once, and cache? or every time?)
4) during backward, the _local_tensor is converted from (initial dtype? or forward dtype) to backward type (once, or every time?)

Relating to (3) and (4), i guess DTensor doesn't do anything to the other 'inputs' to a forward or backward operator that uses this DTensor.  I guess this DTensor's local_tensor is converted to the new type, then along with other DTensor inputs they are fed into the aten op, and the precision matching (or mixed precision) behaviors (and activation output dtype) are defined by aten's type promotion rules?

[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/tensor/_redistribute.py:346]: Hmm this is a good point. In the SimpleFSDP use case, it has always been "initial dtype" = "backward dtype". I think in general we should (1) perform backward redistributed using "backward dtype" (as is done now), and (2) perform another dtype conversion from "backward dtype" back to "initial dtype".

This would also align with FSDP2 mixed precision reduce_dtype behavior.

To answer your questions:
> once, and cache? or every time?

The dtype conversion would happen every time.

> the precision matching (or mixed precision) behaviors (and activation output dtype) are defined by aten's type promotion rules

I didn't get this part. In SimpleFSDP there is no such (implicit?) type conversions.

File: test/distributed/tensor/test_redistribute.py [hunk 3]
```diff
@@ -196,6 +204,85 @@ def test_partial_to_replicate_forward_backward(self):
...
+        replica_tensor = distribute_tensor(local_tensor, device_mesh, replica_spec)
+        with comm_mode:
+            reshard_replica_tensor = replica_tensor.redistribute(
+                device_mesh,
+                replica_spec,
+                forward_dtype=self.forward_dtype,
+                backward_dtype=self.backward_dtype,
+            )
+        self.assertEqual(replica_tensor.size(), local_tensor.size())
+        self.assertEqual(replica_tensor.to(self.forward_dtype), reshard_replica_tensor)
```
[the remark this answers is not in the document]
Comment (ruisizhang123) [reply test/distributed/tensor/test_redistribute.py:226]: actually it didn't pass the test haha. `reshard_replica_tensor` is `bfloat16`, which throws datatype mismatch errors when `reshard_replica_tensor` is `float32`.

The reason is in `replicate_to_replicate` case, the `current_spec.placements` and `placements` [here](https://github.com/pytorch/pytorch/blob/121072634ed247c04688bbae6071d5f5fe8c04db/torch/distributed/tensor/_redistribute.py#L318C39-L318C49) is the same,  and `output` is the casted `local_tensor` in bfloat16 hmmmm

File: test/distributed/tensor/test_redistribute.py [hunk 4]
```diff
@@ -196,6 +204,85 @@ def test_partial_to_replicate_forward_backward(self):
...
+        self.assertEqual(comm_mode.get_total_counts(), 0)
+
+        # 2) test replicate -> replicate backward:
+        # should give gradient as replicate
+        grad_output = torch.ones_like(reshard_replica_tensor)
+        with comm_mode:
+            reshard_replica_tensor.backward(grad_output)
+        grad_input = replica_tensor.grad
+        self.assertEqual(grad_input.placements, replica_spec)
```
[the remark this answers is not in the document]
Comment (ruisizhang123) [reply test/distributed/tensor/test_redistribute.py:235]: sg, I've added more examples & links to SimpleFSDP in the PR description. I'm also adding another set (fwd=f32 and bwd=bf16).

**Review by XilunWu — COMMENTED** [review]
I wonder if this type conversion must happen in `redistribute`. Can this be implemented in a less intrusive way than changing DTensor API (e.g. as a hook)?

File: torch/distributed/tensor/_redistribute.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/tensor/_redistribute.py:346]: @XilunWu with mixed precision there is only one such call within FSDP (for each parameter), please see https://github.com/pytorch/torchtitan/pull/1060

File: torch/distributed/tensor/_api.py [hunk 5]
```diff
@@ -513,6 +515,8 @@ def redistribute(
         Keyword args:
             async_op (bool, optional): whether to perform the DTensor redistribute operation
                 asynchronously or not. Default: False
+            forward_dtype (torch.dtype, optional): in the forward pass, the DTensor's datatype can be converted to forward_dtype. Default: None
+            backward_dtype (torch.dtype, optional): in the backward pass, the DTensor's datatype can be converted to backward_dtype. Default: None
```
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/tensor/_api.py:519]: This PR:
foward: fp32 -> bf16 -> all-gather
backward: bf16 -> fp32 -> reduce-scatter

Your code:
forward: fp32 -> bf16 -> all-gather
backward: bf16 -> reduce-scatter -> fp32

**Comment by tianyu-l** [comment]
@XilunWu 
> I wonder if this type conversion must happen in redistribute. Can this be implemented in a less intrusive way than changing DTensor API (e.g. as a hook)?

Could you think of a concrete proposal? If it works in eager, would the backward hooks make it harder for torch.compile to generate full graph?

**Review by wanchaol — CHANGES_REQUESTED** [review]
Nice feature added! I think the addition to the API make sense, however I think the mixed precision handling for autograd have some issues we need to resolve. Please see inlined comments

**Comment by ruisizhang123** [comment]
> @XilunWu
> 
> > I wonder if this type conversion must happen in redistribute. Can this be implemented in a less intrusive way than changing DTensor API (e.g. as a hook)?
> 
> Could you think of a concrete proposal? If it works in eager, would the backward hooks make it harder for torch.compile to generate full graph?

Ya, iiuc we are avoiding hooks in SimpleFSDP because it is hard to make the order correct in full-graph tracing per this [link](https://pytorch.org/docs/stable/torch.compiler_faq.html#do-you-support-distributed-code
).

File: test/distributed/tensor/test_redistribute.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ruisizhang123) [reply test/distributed/tensor/test_redistribute.py:226]: sg, that's a good point.

File: torch/distributed/tensor/_redistribute.py [hunk 6]
```diff
@@ -292,14 +292,32 @@ def forward(  # type: ignore[override]
         device_mesh: DeviceMesh,
         placements: tuple[Placement, ...],
         async_op: bool = False,
+        forward_dtype: Optional[torch.dtype] = None,
+        backward_dtype: Optional[torch.dtype] = None,
     ):
-        current_spec = input._spec
-        ctx.current_spec = current_spec
         ctx.async_op = async_op
+        ctx.backward_dtype = backward_dtype
```
[the remark this answers is not in the document]
Comment (ruisizhang123) [reply torch/distributed/tensor/_redistribute.py:299]: this is a very good point. for now since it's major use case is SimpleFSDP (which requires both fwd_dtype and bwd_dtype), can we enforce an assertion in API, asking users to specific both fwd_dtype and bwd_dtype, if one of them is not None 🤔️

[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/tensor/_redistribute.py:299]: I think we should always cast back to the original dtype, even if `backward_dtype` is given.
See my comment at https://github.com/pytorch/pytorch/pull/150740#discussion_r2032331579

It is debatable, if `backward_dtype` is not given, should we perform backward redistribute in `forward_dtype` or original dtype. I personally think it should be the latter.

[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/tensor/_redistribute.py:299]: Yeah make sense, I think it should be original dtype, then let's convert back to the original dtype if the local tensor dtype is different.

File: test/distributed/tensor/test_redistribute.py [hunk 7]
```diff
@@ -196,6 +196,97 @@ def test_partial_to_replicate_forward_backward(self):
...
+            grad_output = torch.ones_like(reshard_replica_tensor)
+            with comm_mode:
+                reshard_replica_tensor.backward(grad_output)
+            grad_input = replica_tensor.grad
+            self.assertEqual(grad_input.placements, replica_spec)
+            self.assertEqual(grad_input.to_local(), torch.ones(12, 3))
+            self.assertEqual(comm_mode.get_total_counts(), 0)
+
+    @with_comms
+    def test_shard_to_replicate_forward_backward_datatype_conversion(self):
```
Comment (wanchaol) [inline test/distributed/tensor/test_redistribute.py:239]: Can you also add tests cases to test when `forward_dtype` and `backward_dtype` being specified alone? It's important to guard the behaviors we just discussed.

Comment (ruisizhang123) [reply test/distributed/tensor/test_redistribute.py:239]: I've added when only `forward_dtype` is specified. I'll added the only `backward_dtype` is specified.

File: torch/distributed/tensor/_redistribute.py [hunk 8]
```diff
@@ -343,11 +385,7 @@ def backward(ctx, grad_output: "dtensor.DTensor"):  # type: ignore[override]
         spec = DTensorSpec(
             previous_spec.device_mesh,
             tuple(normalized_placements),
-            tensor_meta=TensorMeta(
-                shape=grad_output.shape,
-                stride=grad_output.stride(),
```
[the remark this answers is not in the document]
Comment (ruisizhang123) [reply torch/distributed/tensor/_redistribute.py:348]: I only added tests for different datatypes and shard with different dims & input sizes. Wonder if you have a pointer to different stride sizes tests that I can add to current test cases?

[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/tensor/_redistribute.py:348]: sorry I don't remember a concrete test case about this, you can probably generate one by hand, also it looks like the CI failures are related

File: torch/distributed/tensor/_redistribute.py [hunk 9]
```diff
@@ -320,17 +339,40 @@ def forward(  # type: ignore[override]
...
+                    stride=grad_output.stride(),
+                    dtype=backward_dtype,
+                ),
+            )
+            previous_spec = DTensorSpec(
+                mesh=previous_spec.device_mesh,
+                placements=previous_spec.placements,
+                tensor_meta=current_spec.tensor_meta,
+            )
+        else:
```
Comment (tianyu-l) [inline torch/distributed/tensor/_redistribute.py:361]: We need to add a branch.
If `backward_dtype` is not given, but `original_dtype != grad_output._local_tensor.dtype` (which should be `forward_dtype` if specified), then we should perform dtype conversion, so that backward redistribute happens in the `original_dtype`.

File: torch/distributed/tensor/_redistribute.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/tensor/_redistribute.py:348]: I forgot the reason why I changed it this way in https://github.com/tianyu-l/pytorch_intern24/pull/20
Let's keep the reconstruction of this `TensorMeta`, but with the right dtype.

**Review by tianyu-l — COMMENTED** [review]
Left inline comments.

Also, please fix linting by `lintrunner init` and `lintrunner -a`.

File: torch/distributed/tensor/_redistribute.py [hunk 10]
```diff
@@ -320,17 +339,42 @@ def forward(  # type: ignore[override]
...
+            if grad_output._local_tensor.dtype != ctx.original_dtype:
+                local_tensor = grad_output._local_tensor.to(ctx.original_dtype)
+            local_tensor = grad_output._local_tensor
+            current_spec = grad_output._spec
```
Comment (tianyu-l) [inline torch/distributed/tensor/_redistribute.py:365]: In the case where `grad_output._local_tensor.dtype != ctx.original_dtype`, we'd still need to create a `TensorMeta` as we do in the first branch. Here's the code I think we can use:
```suggestion
        backward_dtype = ctx.backward_dtype or ctx.original_dtype

        if backward_dtype != grad_output._local_tensor.dtype:
            local_tensor = grad_output._local_tensor.to(dtype=backward_dtype)
            current_spec = DTensorSpec(
                mesh=grad_output._spec.device_mesh,
                placements=grad_output._spec.placements,
                tensor_meta=TensorMeta(
                    shape=grad_output.shape,
                    stride=grad_output.stride(),
                    dtype=backward_dtype,
                ),
            )
            previous_spec = DTensorSpec(
                mesh=previous_spec.device_mesh,
                placements=previous_spec.placements,
                tensor_meta=current_spec.tensor_meta,
            )
        else:
            local_tensor = grad_output._local_tensor
            current_spec = grad_output._spec
```

Comment (ruisizhang123) [reply torch/distributed/tensor/_redistribute.py]: make sense, updated!

File: torch/distributed/tensor/_redistribute.py [hunk 11]
```diff
@@ -292,14 +292,33 @@ def forward(  # type: ignore[override]
...
+        forward_dtype: Optional[torch.dtype] = None,
+        backward_dtype: Optional[torch.dtype] = None,
     ):
-        current_spec = input._spec
-        ctx.current_spec = current_spec
         ctx.async_op = async_op
+        ctx.backward_dtype = backward_dtype
+        ctx.original_dtype = input._local_tensor.dtype
+
+        if forward_dtype is not None:
```
Comment (tianyu-l) [inline torch/distributed/tensor/_redistribute.py:302]:
```suggestion
        if forward_dtype is not None and forward_dtype != input._local_tensor.dtype:
```

**Review by tianyu-l — APPROVED** [review]
LGTM!
